### PR TITLE
WIP: added python script with ekf derivation

### PR DESCRIPTION
**Pull request by RomanBapst**
### Motivation
Some time ago I decided to implement the 24 state EKF filter equations in python for the following reasons:

- I do not have a Matlab license anymore
- I wanted to be able to derive the equations quickly and not wait for 3 hours every time I change something
- I wanted to use this in combination with an editor which allows me to interact with the equations once they were generated

### What's the state of this
So far I have implemented:

- definition of states and state propagation
- derivation of covariance matrix prediction and simplification (most interesting one)
- derivation of 3D mag fusion
- derivation of vel/pos fusion

I used simpy's [cse](https://docs.sympy.org/latest/modules/rewriting.html#common-subexpression-detection-and-collection) for doing common subexpression detection.
My initial observation with this comparing to the brute force algorithm implemented in Matlab is:

- it's very fast, simplifying the covariance matrix update takes between 10 and 20 seconds
- it produces more intermediate variables than the matlab version
- it produces code with less mathematical operations (+, -, *)

### Verification of the equations
In order to test the correctness of the equations I numerically evaluated the predicted covariance matrix using random numbers. I then summed each row of the covariance matrix and compared the values between the two derivations (c++ code generated by matlab and the python code).
The values look exactly the same as seen in the figure below.
So I'm currently assuming that the derivation is mathematically correct and there are no big mistakes in the code anymore. This was just some initial experiment though, this needs to be done properly moving forward.
image

Update:
Compared the executing time of the covariance matrix update between this derivation here and the one from matlab. I ran both versions for 10000 times in a loop side by side and evaluated the performance using calgrind. The resulting graph can be seen below:
image

I also noticed some slight differences in the numerical result of the predicted covariance matrix between the two derivations. The figure below shows two matrices:
1) The difference between the predicted matrices, e.g. `P_next_matlab - P_next_python` multiplied by a factor of 1e6 in order to make the error more visible.
2) The second matrix shows the same error but relative  to `P_next - P` in percent. In other words, "how big are those differences in magnitude compared to the delta you add to `P` in order to arrive at `P_next` "
I wanted to figure out if the small differences seen in 1) are negligible, and it seems like they are.
However, it would still be nice to figure out where those differences come from.
image

**Update**
Converted all floats to double in both derivations and the errors are gone!

### Editor
I was using [spyder](https://www.spyder-ide.org/) which allows you to interact with the script.

fixes https://github.com/PX4/ecl/issues/667

### CODE REVIEW

**Comment by jkflying** [comment]
Awesome! Could you put the two outputs (I guess for the covariance, since that is most performance critical) into functions in C++ files, and run them in callgrind? Then we can see what the actual instruction count differences are.

**Comment by RomanBapst** [comment]
@jkflying Yes, I'm already on this.

**Comment by RomanBapst** [comment]
Update: I cleaned up the code a bit and pushed everything to this branch. If you run main.py you will get autogenerated cpp files in the same directory.

I also tested the covariance prediction on a Pixhawk cube and got the following results:

- perf showed that the ekf2 average execution time dropped slightly but not significantly (850us vs 811us avg) on an F4
- saved about 3.1K of flash on an F4

File: EKF/python/ekf_derivation/code_gen.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            write_string = write_string + identifier + " = " + self.get_ccode(matrix[0]) + ";\n"
+        elif matrix.shape[0] == 1 or matrix.shape[1] == 1:
+            for i in range(0,len(matrix)):
+                write_string = write_string + identifier + "(" + str(i) + ") = " + self.get_ccode(matrix[i]) + ";\n"
+        else:
+            #write_string = "float " + identifier + "[" + str(matrix.shape[0]) + "][" + str(matrix.shape[1]) + "] = {};\n"
+            for j in range(0, matrix.shape[1]):
+                for i in range(0, matrix.shape[0]):
+                    if j >= i or not is_symmetric:
+                        write_string = write_string + identifier + "(" + str(i) + "," + str(j) + ") = " + self.get_ccode(matrix[i,j]) + ";\n"
```
Comment (jkflying) [inline EKF/python/ekf_derivation/code_gen.py:43]: These all need square brackets now, also above
```suggestion
                        write_string = write_string + identifier + "[" + str(i) + "," + str(j) + "] = " + self.get_ccode(matrix[i,j]) + ";\n"
```

**Comment by dusan19** [comment]
> Update
> Converted all floats to double in both derivations and the errors are gone!

The results you got when using doubles (same with both matlab and python equations), are they closer to matlab float results or to python float results?
Assuming the differences come from float precision limitations and that results obtained with doubles are the ground truth, this comparison would tell us if the python equations make the numerical situation better or worse than matlab equations. Hopefully it would just be a verification that python generated equations using floats are closer to ground truth (doubles).

**Comment by priseborough** [comment]
Can you do something about the powf( arg,2) instances in the generated code. If not we can do a find and replace with sq(arg) in the editor.

**Comment by bresch** [comment]
@RomanBapst Could you please follow up on this?

**Comment by dusan19** [comment]
I managed to create the log from boot where the estimator crashes every 15 min for 2hrs, and replayed it with the changes from this PR.
here are the logs of the original log replay, and the replay with this PR, respectively.
The logs show that the python generated equations do not solve the problem, the estimator still crashes, but the crashing is less often, first time it happened after more than an hour from boot

Original:

orig_replay

With python generated equations:

python_cov

**Comment by priseborough** [comment]
This concept has been developed further by https://github.com/PX4/ecl/pull/870 and subsequent PR's

**Pull request closed without merging**